ssl/quic/quic_channel.c

```diff
@@ -113,6 +114,88 @@ static int gen_rand_conn_id(OSSL_LIB_CTX *libctx, size_t len, QUIC_CONN_ID *cid)
     return 1;
 }
 
+static unsigned long chan_reset_token_hash(
+        const QUIC_STATELESS_RESET_TOKEN *tok)
+{
+    unsigned long h;
+
+    assert(sizeof(h) <= sizeof(tok->token));
+    memcpy(&h, tok->token, sizeof(h));
+    return h;
+}
+
+static int chan_reset_token_cmp(const QUIC_STATELESS_RESET_TOKEN *t1,
+                                const QUIC_STATELESS_RESET_TOKEN *t2)
+{
+    return CRYPTO_memcmp(t1->token, t2->token, sizeof(t1->token));
+}
+
+static int reset_token_obfuscate(QUIC_STATELESS_RESET_TOKEN *out,
+                                 const unsigned char *in, size_t len)
+{
+    /*
+     * TODO(QUIC): update this to AES encrypt the token in ECB mode with a
+     * random (per channel) key.
+     */
+    memcpy(out, in, sizeof(*out));
+    return 1;
+}
+
+/*
+ * Remove the old token from the table of legal tokens and insert the new
+ * one in it's place.
+ */
+static int chan_update_reset_token(QUIC_CHANNEL *ch,
+                                   const unsigned char *new, size_t new_len)
+{
+    QUIC_STATELESS_RESET_TOKEN *r;
+
+    /*
+     * Remove any existing reset token, if this fails it means there wasn't
+     * one and this can safely be ignored.
+     */
+    (void)lh_QUIC_STATELESS_RESET_TOKEN_delete(ch->stateless_reset_tokens,
+                                               &ch->cur_remote_reset_token);
+    reset_token_obfuscate(&ch->cur_remote_reset_token, new, new_len);
+    r = lh_QUIC_STATELESS_RESET_TOKEN_insert(ch->stateless_reset_tokens,
+                                             &ch->cur_remote_reset_token);
+    return r != NULL;
+}
+
+/*
+ * This is called by the demux whenever a new datagram arrives
+ */
+static int ch_stateless_reset_token_handler(const unsigned char *data,
+                                            size_t datalen, void *arg)
+{
+    QUIC_STATELESS_RESET_TOKEN tok;
+    QUIC_CHANNEL *ch = (QUIC_CHANNEL *)arg;
+
+    /*
+     * Perform some fast and cheap checks for a packet not being a stateless
+     * reset token.  RFC 9000 s. 10.3 specifies this layout for stateless
+     * reset packets:
+     *
+     *  Stateless Reset {
+     *      Fixed Bits (2) = 1,
+     *      Unpredictable Bits (38..),
+     *      Stateless Reset Token (128),
+     *  }
+     *
+     * We can rapidly check for the minimum length and that the first pair
+     * of bits in the first byte are 01.
+     *
+     * The funtion returns 1 if it is a stateless reset packet, 0 if it isn't
+     * and -1 if an error was encountered.
+     */
+    if (datalen < QUIC_STATELESS_RESET_TOKEN_LEN + 5 || (0300 & *data) != 0100)
+        return 0;
+    if (!reset_token_obfuscate(&tok, data, datalen))
+        return -1;
+    return lh_QUIC_STATELESS_RESET_TOKEN_retrieve(ch->stateless_reset_tokens,
```

Comment:
So I guess there's an implicit TODO here on actually handling the stateless reset event. OK.

Reply:
I think it **is** handled by the caller correctly.  Returning true here causes `demux_process_pending_urxe` to return failure immediately.  This leads to `ossl_quic_demux_pump` returning `QUIC_DEMUX_PUMP_RES_PERMANENT_FAIL` skips to the terminating state without sending a close frame.

Serendipity that the code path was already there.

Reply:
I mean... okay. But we won't have an informative termination reason set or anything, and the QUIC error code is probably wrong.

I'd really prefer to see this handled explicitly with correct termination information.

Annoyingly, there is no QUIC error code assigned which would be suitable here. We could assign our own value greater than `1<<62` for reporting to the application. Hmm..

Reply:
Isn't the stateless reset equivalent of connection close with QUIC_ERR_NO_ERROR code?